changed parks to all pois

mortonrj opened the pull request.

Commit:
added all pois

Commit:
added helper functions + multiple types + fixed bugs

Commit:
cleaning up

mortonrj committed:
cleaning up

mortonrj committed:
Merge branch 'rachael' of https://github.com/ariadnes-thread/ariadne-route-planner into rachael

mortonrj committed:
bug fixes

Commit:
Only add requirements that are used

squarespirit commented:
Hey Rachael, I pushed a commit that updates requirements.txt. Next time please include only the ones that are used.

Note that one of the outputs of `make_route` is some GeoJSON. Consider visualizing it on a site like http://geojsonlint.com/ to make sure it goes through the POIs you expect.

When your functions have member data, consider wrapping them in a class. For example, my OrienteeringRouter class has a `gmapsclient` and database connection `conn` as member data. Otherwise those variables become global variables. It makes sense that your `google_utils.py` needs a Google API key to do any of the requests. So consider wrapping it in a class. My OrienteeringRouter class also takes the API key and DB connection as parameters instead of fetching them itself. That way, it makes the least assumptions about the world.

For the different location types, note that the location type a user enters on the frontend may be _different_ from the name of the type used by the GMaps API (though the result from the frontend will be a string), so there might need to be some conversion that happens. I'm not really sure what the right thing to do is yet, let me know if we should talk about it!

squarespirit commented:
Also, Pycharm is a great IDE for Python! It will show all your syntax errors. Lmk if you need help setting it up

TimboKZ commented:
Matt has been converted

veniversum commented on planner/utils/google_utils.py:
```diff
@@ -0,0 +1,87 @@
+
+	return query_result.rating
+
+def save_to_db():
```
This function doesn't do anything. It just opens a new connection and closes it immediately.

veniversum commented on planner/utils/google_utils.py:
```diff
@@ -0,0 +1,87 @@
+import db_conn
+import json
+
+config = ConfigParser()
```
`ConfigParser` instance is never used.

veniversum commented on planner/orienteering_router.py:
```diff
@@ -1,5 +1,7 @@
 from collections import namedtuple
 import json
+import sys
```
Remove unused import.

veniversum commented on planner/orienteering_router.py:
```diff
@@ -8,6 +10,7 @@
 import db_conn
```
This isn't introduced in this changeset, but this import should be moved to within `main()` or better, inside 
```if __name__ == '__main__':```, along with the definition of `main()`.

We don't want to create a new `ThreadedConnectionPool` if this module is imported, but doesn't need database connections..

veniversum commented on planner/utils/google_utils.py:
```diff
@@ -0,0 +1,87 @@
+https://github.com/slimkrazy/python-google-places
+"""
+
+from googleplaces import GooglePlaces, types, lang
```
Unused import `lang`

veniversum commented on planner/orienteering_router.py:
```diff
@@ -27,38 +30,37 @@ def __init__(self, gmaps_api_key: str, conn):
-            type='park'
-        )
-
+        res = GoogleUtils.get_pois({'lat': location[0], 'lng': location[1]}, radius=radius, type_list=['park'])
```
Is this intended as `type_list=type_list`?

veniversum commented on planner/utils/google_utils.py:
```diff
@@ -0,0 +1,87 @@
+	"""
+	Gets ratings for a point of interest, returns json
+	"""
+	query_result = google_places.get_place(place_id)
```
`place_id` needs to be passed as an arg.

veniversum commented on planner/utils/google_utils.py:
```diff
@@ -0,0 +1,87 @@
+	conn.commit()
+	cur.close()
+	conn.close()
+
```
This file is indented using tabs instead of spaces.

veniversum commented on planner/orienteering_router.py:
```diff
@@ -27,38 +30,37 @@ def __init__(self, gmaps_api_key: str, conn):
         single API request, and only returns up to 20 places.
         :param location: (lat, lon) pair.
         :param radius: Search radius in meters.
+        :param types: List of the type of POIS being fetched from gmaps
```
`type_list` instead of `types`.

veniversum commented on planner/orienteering_router.py:
```diff
@@ -27,38 +30,37 @@ def __init__(self, gmaps_api_key: str, conn):
-    def get_parks_from_gmaps(self, location: Tuple[float, float],
-                             radius: float) -> List[GmapsResult]:
+    def get_pois_from_gmaps(self, location: Tuple[float, float],
+                             radius: float, type_list=['park']) -> List[GmapsResult]:
```
Default arg `['park']` is mutable.

https://docs.quantifiedcode.com/python-anti-patterns/correctness/mutable_default_value_as_argument.html

veniversum commented on planner/utils/google_utils.py:
```diff
@@ -0,0 +1,87 @@
+		json.dump(data, outfile)
+
+
+def get_pois(lat_lng, radius=15, type_list=['park']):
```
Default arg `['park']` is mutable.

veniversum reviewed (changes requested):
@squarespirit too, whenever calling `db_conn.connPool.getconn()`, make sure to return the connection to the pool, if not the connection pool will get depleted.

veniversum commented:
Sorry, some of these were from #1 which I missed. We could fix as many as we can and refactor the rest later.

We should discuss the design of the route planner during the next meeting, including decoupling the planner from its dependency on Gmaps for POIs so that we can swap out the POI provider (to Yelp, our own DB, or others) easily.

Also, it feels like most of the methods in `OrienteeringRouter` can be made static, we don't really need/want to create a new instance per request.

squarespirit commented:
Sorry, would it be possible for RoutePlanner in start_server.py to create only one OrienteeringRouter and use that for everything?

Agree that we should discuss design

squarespirit commented:
types.py has the same name as a Python builtin module https://docs.python.org/3.7/library/types.html. This was causing problems when I was running things in Pycharm.

mortonrj committed:
gerge branch 'master' of https://github.com/ariadnes-thread/ariadne-route-planner into rachael

mortonrj committed:
Cleaned up + added class to google_utils file. Added general_utils file with conversions between zip code, latitude + longitude, and address

mortonrj committed:
Merge branch 'rachael' of https://github.com/ariadnes-thread/ariadne-route-planner into rachael

mortonrj committed:
updated requirements.txt

mortonrj committed:
small fix

mortonrj committed:
fixed some things I forgot to

mortonrj committed:
fixed some things I forgot to

mortonrj commented:
I made the suggested changes. Could you guys look them over? I want to commit a few more things but I think I should do it in a different pull request. @veniversum @squarespirit

veniversum commented on planner/utils/google_utils.py:
```diff
@@ -0,0 +1,79 @@
+	def __init__(self, gmaps_api_key=None, conn=None):
+		import db_conn
+		if gmaps_api_key is None:
+			config = ConfigParser()
```
`ConfigParser` here is unused. If it's unneeded we can remove it and the import on Line 12 too.

mortonrj replied:
Oops - I'm going to just take all the db stuff out in this file because it's not being used as of now

veniversum commented on planner/orienteering_router.py:
```diff
@@ -27,38 +28,38 @@ def __init__(self, gmaps_api_key: str, conn):
-        )
-
+        GoogleHelper = GoogleUtils.GoogleHelper() 
+        res = GoogleHelper.get_pois({'lat': location[0], 'lng': location[1]}, radius=radius, type_list=['park'])
```
The keyword arg `type_list` should use the passed in `type_list` arg instead of `['park']` right?

If not the passed in arg `type_list` is unused.

mortonrj replied:
Changed the input type

veniversum commented on planner/utils/google_utils.py:
```diff
@@ -0,0 +1,79 @@
+		"""
+		Gets ratings for a point of interest, returns json
+		"""
+		query_result = self.google_places.get_place(place_id)
```
You need to pass in `place_id` as an arg, it's referenced without being declared.

veniversum commented on planner/utils/google_utils.py:
```diff
@@ -0,0 +1,79 @@
+			json.dump(data, outfile)
+
+
+	def get_pois(self, lat_lng, radius=15, type_list=['park']):
```
Default arg `['park']` is mutable. It's harmless now but might lead to bugs in the future.

docs.quantifiedcode.com/python-anti-patterns/correctness/mutable_default_value_as_argument.html

veniversum commented on planner/utils/google_utils.py:
```diff
@@ -0,0 +1,79 @@
+			    config = json.load(f)
+			gmaps_api_key = config["gmapsApiKey"]
+		if conn is None:
+			conn = db_conn.connPool.getconn()
```
This leads to connections being eventually exhausted from the connection pool. Should return connection when done.

Consider implementing `__enter__` and `__exit__` in order to support the `with GoogleHelper() as ghelper:` syntax and cleans up after itself.

veniversum reviewed (changes requested):
Some unresolved issues from the previous review.

`GoogleHelper` causes DB connection pool to deplete the way it's currently implemented/used.

squarespirit commented on planner/utils/types.py:
```diff
@@ -0,0 +1,143 @@
+"""
```
Please change the name of this file, to e.g. `poi_types.py`? It currently conflicts with a Python builtin module https://docs.python.org/3/library/types.html.

squarespirit commented on planner/orienteering_router.py:
```diff
@@ -27,38 +28,38 @@ def __init__(self, gmaps_api_key: str, conn):
             except KeyError:
                 # Skip POIs that are missing one of the fields
                 pass
 
```
This function wasn't returning any POIs, because each POI threw a KeyError in this block. I think the key for the longitude is supposed to be 'lng', and the try/except block might not be needed anymore.

mortonrj replied:
Oops - fixed

mortonrj committed:
made some fixes

mortonrj commented:
I think I updated with the request changes. Let me know if you guys have any more input!

veniversum commented on planner/utils/google_utils.py:
```diff
@@ -0,0 +1,62 @@
+			    config = json.load(f)
+			gmaps_api_key = config["gmapsApiKey"]
+
+		self.google_places = GooglePlaces(config["gmapsApiKey"])
```
Not using `gmaps_api_key` passed in.

veniversum committed:
Fix remaining issues in #2

veniversum commented:
Changed base branch to `develop` since `master` pushes to production.

Fixed the remaining of the outstanding issues, @squarespirit you can merge after you sign off if there aren't any remaining issues.

Commit:
Skip GMap POIs without ratings

Commit:
Use 'type' instead of 'types' in GMaps search

'types' param seems to be deprecated
https://github.com/slimkrazy/python-google-places
and is not even mentioned on Google's API documentation anymore
https://developers.google.com/places/web-service/search.
'type' seems to return higher quality results. For example,
https://maps.googleapis.com/maps/api/place/nearbysearch/json?location=34.140003,-118.122775&radius=3000&types=park|restaurant&key=YOUR_API_KEY
returns mainly parks, even though 'types' contains both park & restuarant.
Individual queries with 'type=park' and 'type=restaurant' are pretty
accurate.  So it may be better to issue multiple nearby search queries,
one for each type.

squarespirit commented:
@mortonrj I changed get_pois to iterate over types and issue a query for each type. The reason is that the 'types' param seems to be deprecated https://github.com/slimkrazy/python-google-places and is not mentioned on Google's API documentation anymore https://developers.google.com/places/web-service/search. 'type' seems to return higher quality results. For example,
```
https://maps.googleapis.com/maps/api/place/nearbysearch/json?location=34.140003,-118.122775&radius=3000&types=park|restaurant&key=YOUR_API_KEY
```
returns mainly parks, even though 'types' contains both park & restuarant. Individual queries with 'type=park' and 'type=restaurant' are pretty accurate.  So it may be better to issue multiple nearby search queries, one for each type.
